src/main/java/net/glowstone/block/itemtype/ItemBow.java

```diff
@@ -0,0 +1,127 @@
+package net.glowstone.block.itemtype;
+
+import java.util.Optional;
+import java.util.concurrent.ThreadLocalRandom;
+import net.glowstone.entity.GlowPlayer;
+import net.glowstone.entity.projectile.GlowTippedArrow;
+import net.glowstone.inventory.GlowInventorySlot;
+import net.glowstone.util.InventoryUtil;
+import org.bukkit.GameMode;
+import org.bukkit.Material;
+import org.bukkit.enchantments.Enchantment;
+import org.bukkit.entity.Arrow;
+import org.bukkit.entity.SpectralArrow;
+import org.bukkit.entity.TippedArrow;
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.inventory.meta.PotionMeta;
+
+public class ItemBow extends ItemTimedUsage {
+    private static final long TICKS_TO_FULLY_CHARGE = 20;
+    private static final double MAX_BASE_DAMAGE = 9;
+    private static final double MAX_SPEED = 53;
+
+    @Override
+    public void startUse(GlowPlayer player, ItemStack item) {
+        if (player.getGameMode() == GameMode.CREATIVE || findArrow(player).isPresent()) {
+            player.setUsageItem(item);
+            player.setUsageTime(TICKS_TO_FULLY_CHARGE);
+        }
+    }
+
+    @Override
+    public void endUse(GlowPlayer player, ItemStack bow) {
+        // Check arrows again, since plugins may have changed the inventory while the bow was drawn
+        Optional<GlowInventorySlot> maybeArrow = findArrow(player);
+        GlowInventorySlot slot = null;
+        ItemStack arrow = null;
+        Material arrowType = Material.AIR;
+        Arrow launchedArrow = null;
+        boolean consumeArrow = false;
+        if (maybeArrow.isPresent()) {
+            slot = maybeArrow.get();
+            arrow = slot.getItem();
+            arrowType = arrow.getType();
+            consumeArrow = (player.getGameMode() != GameMode.CREATIVE);
+        } else if (player.getGameMode() == GameMode.CREATIVE) {
+            // Can fire without arrows in Creative
+            arrowType = Material.ARROW;
+            consumeArrow = false;
+        }
+        switch (arrowType) {
+            case ARROW:
+                if (bow.containsEnchantment(Enchantment.ARROW_INFINITE)) {
+                    consumeArrow = false;
+                }
+                launchedArrow = player.launchProjectile(Arrow.class);
+                break;
+            case TIPPED_ARROW:
+                launchedArrow = player.launchProjectile(TippedArrow.class);
+                GlowTippedArrow launchedTippedArrow = (GlowTippedArrow) launchedArrow;
+                launchedTippedArrow.copyFrom((PotionMeta) arrow.getItemMeta());
+                break;
+            case SPECTRAL_ARROW:
+                launchedArrow = player.launchProjectile(SpectralArrow.class);
+                break;
+            case AIR:
+                // Not in creative mode and have no arrow
+                break;
+            default:
+                player.getServer().getLogger()
```

Comment:
`GlowServer.logger`

Reply:
Done in f656da22b66794f0e340518025296d49d53a43a8 (also migrated to the `Supplier<String>` logging function).